### PR TITLE
Fix Linux packages

### PR DESCRIPTION
**Pull request by daxian-dbw**
Part of the work for #3961

`dotnet-runtime-2.0.0` Linux packages for Debian distributions depend on different lib packages on different distros, for example, `libicu52` on `ubuntu.14.04`, `libicu55` on `ubuntu.16.04`, `libicu57` on `ubuntu.17.04`, `libicu52` on `debian 8`, and `libicu57` on `debian 9`. A similar issue also happens to `libssl` -- `libssl1.0.2` on `debian 9` and `libssl1.0.0` on the rest debian distros.

In order to use a single DEB package for all distros, we need to depend on `-dev` packages, like `libicu-dev` and `libssl-dev`. However, dev packages are not meant for production software, and dotnet core has separate DEB package for each Debian distro, so I think we should do that too.

So for Debian distros, we have a separate package for each distro. For Red Hat family, we have a single RPM package, just like the `dotnet-runtime-2.0.0` package.

Even though we have separate packages for Debian distros, we now can create them all from one single build. Given that AppImage current needs to be created on `ubuntu.14.04`, we can generate all pacakges from `ubuntu.14.04`.
For RPM package, we should generate it from `CentOS7`.

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@TravisEz13 @adityapatwardhan The changes in `Start-PSPackage` looks huge, but actually is not that much. It shows a big chunk of changes because the existing script needs to be moved to `End { ... }` block due to the introduction of the dynamic parameter `-Distribution`. This dynamic parameter is enabled only if `-Type deb` is specified, so that we can indicate which Debian distro the package is targeting.

**Comment by daxian-dbw** [comment]
I have tested all packages on all supported distros except for Red Hat Enterprise Linux 7. They are:
`Ubuntu.14.04`, `Ubuntu.16.04`, `Ubuntu.17.04`, `CentOS7`, `openSUSE 42,2`, `Fedora 25`, `Fedora 26`.

File: tools/packaging/packaging.psm1 [hunk 1]
```diff
@@ -246,255 +260,305 @@ function New-UnixPackage {
...
+                "libgcc1",
+                "libgssapi-krb5-2",
+                "liblttng-ust0",
+                "libstdc++6",
+                "libunwind8",
+                "libuuid1",
+                "zlib1g"
+            )
+
+            if ($DebDistro -eq "ubuntu.14.04") {
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1]: would this be more readable as a switch statement?

File: build.psm1 [hunk 2]
```diff
@@ -131,9 +131,13 @@ function Get-EnvironmentInformation
         $LinuxInfo = Get-Content /etc/os-release -Raw | ConvertFrom-StringData
 
         $environment += @{'LinuxInfo' = $LinuxInfo}
+        $environment += @{'IsDebian' = $LinuxInfo.ID -match 'debian'}
+        $environment += @{'IsDebian8' = $Environment.IsDebian -and $LinuxInfo.VERSION_ID -match '8'}
+        $environment += @{'IsDebian9' = $Environment.IsDebian -and $LinuxInfo.VERSION_ID -match '9'}
         $environment += @{'IsUbuntu' = $LinuxInfo.ID -match 'ubuntu'}
         $environment += @{'IsUbuntu14' = $Environment.IsUbuntu -and $LinuxInfo.VERSION_ID -match '14.04'}
         $environment += @{'IsUbuntu16' = $Environment.IsUbuntu -and $LinuxInfo.VERSION_ID -match '16.04'}
+        $environment += @{'IsUbuntu17' = $Environment.IsUbuntu -and $LinuxInfo.VERSION_ID -match '17.04'}
```
Comment (TravisEz13) [inline build.psm1]: build.psm1 LGTM

File: tools/appimage.sh [hunk 3]
```diff
@@ -409,7 +409,7 @@ OPTIONS="-o Debug::NoLocking=1
 -o APT::Install-Suggests=0
 "
 
-cp ../powershell_*_amd64.deb .
+cp ../powershell_*ubuntu.14.04_amd64.deb .
```
Comment (TravisEz13) [inline tools/appimage.sh]: appimage.sh LGTM

**Review by TravisEz13 — APPROVED** [review]
One minor comment

File: tools/packaging/packaging.psm1 (hunk 1 above)
Comment (daxian-dbw) [reply tools/packaging/packaging.psm1]: possible. I will use `switch` instead.

File: tools/packaging/packaging.psm1 [hunk 4]
```diff
@@ -246,255 +260,301 @@ function New-UnixPackage {
...
-        }
-        "osxpkg" {
-            if (!$Environment.IsMacOS) {
-                throw ($ErrorMessage -f "macOS")
-            }
+    DynamicParam {
+        if ($Type -eq "deb") {
+            # Add a dynamic parameter '-Distribution' when the specified package type is 'deb'.
+            # The '-Distribution' parameter can be used to indicate which Debian distro this pacakge is targeting.
+            $ParameterAttr = New-Object "System.Management.Automation.ParameterAttribute"
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1]: use ```[System.Management.Automation.ParameterAttribute]::new()``` instead

File: tools/packaging/packaging.psm1 [hunk 5]
```diff
@@ -246,255 +260,301 @@ function New-UnixPackage {
...
+            $ValidateSetAttr = New-Object "System.Management.Automation.ValidateSetAttribute" -ArgumentList $Script:DebianDistributions
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1]: Use ```[System.Management.Automation.ValidateSetAttribute]::new($DebianDistributions)``` instead

File: tools/packaging/packaging.psm1 [hunk 6]
```diff
@@ -246,255 +260,301 @@ function New-UnixPackage {
...
+            $Attributes = New-Object "System.Collections.ObjectModel.Collection``1[System.Attribute]"
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1]: same as above.

File: tools/packaging/packaging.psm1 [hunk 7]
```diff
@@ -246,255 +260,301 @@ function New-UnixPackage {
...
+            $Attributes.Add($ParameterAttr) > $null
+            $Attributes.Add($ValidateSetAttr) > $null
+
+            $Parameter = New-Object "System.Management.Automation.RuntimeDefinedParameter" -ArgumentList ("Distribution", [string], $Attributes)
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1]: Same as above

File: tools/packaging/packaging.psm1 [hunk 8]
```diff
@@ -246,255 +260,301 @@ function New-UnixPackage {
...
+            $Dict = New-Object "System.Management.Automation.RuntimeDefinedParameterDictionary"
```
Comment (adityapatwardhan) [inline tools/packaging/packaging.psm1]: Same as above.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: tools/packaging/packaging.psm1 (hunk 4 above)
Comment (daxian-dbw) [reply tools/packaging/packaging.psm1]: We use `New-Object` in build.psm1, and thus I used `New-Object` here to keep consistent.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@TravisEz13 @adityapatwardhan Thank you for the review!

**Pull request merged**